From 1a4884245444b9ba62daf412183440e586de8863 Mon Sep 17 00:00:00 2001
From: saikumaru <saikumaru@ami.com>
Date: Tue, 9 Sep 2025 10:49:43 +0530
Subject: [PATCH] Implement hs401x driver

This driver implements support for the Renesas hs401x chip,
a humidity and temperature sensor.

Signed-off-by: saikumaru <saikumaru@ami.com>
---
 drivers/hwmon/Kconfig  |  11 ++
 drivers/hwmon/Makefile |   1 +
 drivers/hwmon/hs401x.c | 253 +++++++++++++++++++++++++++++++++++++++++
 3 files changed, 265 insertions(+)
 create mode 100644 drivers/hwmon/hs401x.c

diff --git a/drivers/hwmon/Kconfig b/drivers/hwmon/Kconfig
index c8d56c42b1d1..78cfab266a6f 100644
--- a/drivers/hwmon/Kconfig
+++ b/drivers/hwmon/Kconfig
@@ -758,6 +758,17 @@ config SENSORS_HDC302X
 	  This driver can also be built as a module. If so, the module
 	  will be called hdc302x.
 
+config SENSORS_HS401X
+	tristate "Renesas HS401x and compatibles"
+	depends on I2C
+	select CRC8
+	help
+	  If you say yes here you get support for Renesas HS4011 and HS4013
+	  relative humidity and temperature sensor chips.
+
+	  This driver can also be built as a module. If so, the module
+	  will be called hs401x.
+
 config SENSORS_HIH6130
 	tristate "Honeywell Humidicon HIH-6130 humidity/temperature sensor"
 	depends on I2C
diff --git a/drivers/hwmon/Makefile b/drivers/hwmon/Makefile
index 69e94adeeb2a..3e901476ccff 100644
--- a/drivers/hwmon/Makefile
+++ b/drivers/hwmon/Makefile
@@ -88,6 +88,7 @@ obj-$(CONFIG_SENSORS_GSC)	+= gsc-hwmon.o
 obj-$(CONFIG_SENSORS_GPIO_FAN)	+= gpio-fan.o
 obj-$(CONFIG_SENSORS_GXP_FAN_CTRL) += gxp-fan-ctrl.o
 obj-$(CONFIG_SENSORS_HDC302X)	+= hdc302x.o
+obj-$(CONFIG_SENSORS_HS401X)   += hs401x.o
 obj-$(CONFIG_SENSORS_HIH6130)	+= hih6130.o
 obj-$(CONFIG_SENSORS_HS3001)	+= hs3001.o
 obj-$(CONFIG_SENSORS_ULTRA45)	+= ultra45_env.o
diff --git a/drivers/hwmon/hs401x.c b/drivers/hwmon/hs401x.c
new file mode 100644
index 000000000000..1dd05ad21378
--- /dev/null
+++ b/drivers/hwmon/hs401x.c
@@ -0,0 +1,253 @@
+/*
+ * Driver for the Renesas HS401x
+ * relative humidity and temperature sensor.
+ */
+
+#include <linux/crc8.h>
+#include <linux/delay.h>
+#include <linux/hwmon.h>
+#include <linux/i2c.h>
+#include <linux/jiffies.h>
+#include <linux/module.h>
+
+/* Measurement Conversion Times (in microseconds) */
+#define HS401X_MEASUREMENT_TIME_ALL_14 1700
+
+/* Command codes */
+#define HS401X_CMD_MEASURE_ALL_HOLD 0xe5
+#define HS401X_CMD_MEASURE_ALL_NO_HOLD 0xf5
+
+#define HS401X_CMD_LEN 1
+#define HS401X_WORD_LEN 4
+#define HS401X_RESP_LEN 5
+#define HS401X_CRC8_POLYNOMIAL 0x1d
+
+DECLARE_CRC8_TABLE(hs401x_crc8_table);
+
+struct hs401x_data {
+	struct i2c_client *client;
+	struct mutex update_lock;
+	bool valid; /* true if following fields are valid */
+	unsigned long last_updated; /* In jiffies */
+	s32 temperature;
+	u32 humidity;
+};
+
+static int hs401x_read_values(struct hs401x_data *data)
+{
+	u8 crc;
+	u16 t_raw, rh_raw;
+	int ret = 0, retry = 3;
+	u8 cmd[HS401X_CMD_LEN] = { HS401X_CMD_MEASURE_ALL_NO_HOLD };
+	u8 raw_data[HS401X_RESP_LEN];
+	struct i2c_client *client = data->client;
+
+	mutex_lock(&data->update_lock);
+
+	if (time_after(jiffies, data->last_updated + HZ) || !data->valid) {
+		ret = i2c_master_send(client, cmd, HS401X_CMD_LEN);
+		if (ret < 0)
+			goto err_unlock;
+retry_read:
+		udelay(HS401X_MEASUREMENT_TIME_ALL_14);
+
+		ret = i2c_master_recv(client, raw_data, HS401X_RESP_LEN);
+		if (ret != HS401X_RESP_LEN) {
+			if(retry-- > 0)
+				goto retry_read;
+			if (ret >= 0)
+				ret = -ENODATA;
+			goto err_unlock;
+		}
+
+		rh_raw = raw_data[0];
+		rh_raw = rh_raw << 8 | raw_data[1];
+		rh_raw = rh_raw & 0x3fff;
+		t_raw = raw_data[2];
+		t_raw = t_raw << 8 | raw_data[3];
+		t_raw = t_raw & 0x3fff;
+
+		crc = crc8(hs401x_crc8_table, &raw_data[0], HS401X_WORD_LEN,
+			   CRC8_INIT_VALUE);
+		if (crc != raw_data[4]) {
+			dev_err(&client->dev, "Measurement data corrupted\n");
+			ret = -EIO;
+			goto err_unlock;
+		}
+
+		/*
+		 * Link of the data sheet (Rev. 1.03):
+		 * https://www.renesas.com/us/en/document/dst/hs40xx-datasheet
+		 *
+		 * As stated in section 7.4.3 (p. 13) of the data sheet,
+		 * the temperature conversion formula is:
+		 * [t_raw / (2^14 - 1)] * 165 - 40 = t_real (in degrees Celsius)
+		 *
+		 * According to the kernel documentation, the unit of the temperature
+		 * is millidegree Celsius, and we have to convert the unit before
+		 * calculating the real input value to prevent precision loss, so
+		 * the formula becomes:
+		 * [t_raw / (2^14 - 1) * 165 - 40] * 1000
+		 * = (t_raw * 165 * 10^3) / (2^14 - 1) - 40000
+		 * ~= (t_raw * 165 * 10^3) / 2^14 - 40000
+		 * = (t_raw * 165 * 5^3) / 2^11 - 40000
+		 * = (t_raw * 20625) / 2^11 - 40000
+		 * = t_real (in millidegrees Celsius)
+		 *
+		 * Similarly, the relative humidity conversion formula stated in the
+		 * data sheet is:
+		 * rh_raw / (2^14 - 1) * 100 = rh_real (in percent)
+		 * After converting to milli-percent, the formula becomes:
+		 * [rh_raw / (2^14 - 1) * 100] * 1000
+		 * = (rh_raw * 10^5) / (2^14 - 1)
+		 * ~= (rh_raw * 10^5) / 2^14
+		 * = (rh_raw * 5^5) / 2^9
+		 * = (rh_raw * 3125) / 2^9
+		 * = rh_real (in milli-percent)
+		 */
+		data->temperature = ((u32)t_raw * 20625 >> 11) - 40000;
+		data->humidity = ((u32)rh_raw * 3125) >> 9;
+		data->last_updated = jiffies;
+		data->valid = true;
+		ret = 0;
+	}
+
+err_unlock:
+	mutex_unlock(&data->update_lock);
+	return ret;
+}
+
+/* hs401x_temperature_read() - read the temperature in millidegrees */
+static int hs401x_temperature_read(struct hs401x_data *data, long *val)
+{
+	int ret;
+
+	ret = hs401x_read_values(data);
+	if (ret < 0)
+		return ret;
+
+	*val = data->temperature;
+
+	return 0;
+}
+
+/* hs401x_humidity_read() - read a relative humidity in millipercent */
+static int hs401x_humidity_read(struct hs401x_data *data, long *val)
+{
+	int ret;
+
+	ret = hs401x_read_values(data);
+	if (ret < 0)
+		return ret;
+
+	*val = data->humidity;
+
+	return 0;
+}
+
+static umode_t hs401x_hwmon_visible(const void *data,
+				   enum hwmon_sensor_types type,
+				   u32 attr, int channel)
+{
+	switch (type) {
+	case hwmon_temp:
+	case hwmon_humidity:
+		return 0444;
+	default:
+		return 0;
+	}
+}
+
+static int hs401x_hwmon_read(struct device *dev, enum hwmon_sensor_types type,
+			    u32 attr, int channel, long *val)
+{
+	struct hs401x_data *data = dev_get_drvdata(dev);
+
+	switch (type) {
+	case hwmon_temp:
+		return hs401x_temperature_read(data, val);
+	case hwmon_humidity:
+		return hs401x_humidity_read(data, val);
+	default:
+		return -EOPNOTSUPP;
+	}
+}
+
+static const struct hwmon_channel_info *hs401x_info[] = {
+	HWMON_CHANNEL_INFO(temp, HWMON_T_INPUT),
+	HWMON_CHANNEL_INFO(humidity, HWMON_H_INPUT),
+	NULL,
+};
+
+static const struct hwmon_ops hs401x_hwmon_ops = {
+	.is_visible = hs401x_hwmon_visible,
+	.read = hs401x_hwmon_read,
+};
+
+static const struct hwmon_chip_info hs401x_chip_info = {
+	.ops = &hs401x_hwmon_ops,
+	.info = hs401x_info,
+};
+
+static int hs401x_probe(struct i2c_client *client)
+{
+	struct device *device = &client->dev;
+	struct device *hwmon_dev;
+	struct hs401x_data *data;
+	int ret;
+
+	ret = i2c_smbus_read_byte_data(client, HS401X_CMD_MEASURE_ALL_HOLD);
+	if (ret < 0) {
+		dev_err(&client->dev, "Failed to register hs401x\n");
+		return ret;
+	}
+
+	data = devm_kzalloc(device, sizeof(*data), GFP_KERNEL);
+	if (!data)
+		return -ENOMEM;
+
+	data->client = client;
+	mutex_init(&data->update_lock);
+
+	crc8_populate_msb(hs401x_crc8_table, HS401X_CRC8_POLYNOMIAL);
+
+	hwmon_dev = devm_hwmon_device_register_with_info(device,
+							 client->name,
+							 data,
+							 &hs401x_chip_info,
+							 NULL);
+
+	if (IS_ERR(hwmon_dev))
+		return PTR_ERR(hwmon_dev);
+
+	dev_info(device, "Found %s chip\n", client->name);
+
+	return 0;
+}
+
+static const struct i2c_device_id hs401x_id[] = {
+	{ "hs401x", 0 },
+	{ },
+};
+MODULE_DEVICE_TABLE(i2c, hs401x_id);
+
+static const struct of_device_id hs401x_of_match[] = {
+	{ .compatible = "renesas,hs401x" },
+	{ }
+};
+MODULE_DEVICE_TABLE(of, hs401x_of_match);
+
+static struct i2c_driver hs401x_driver = {
+	.driver = {
+		.name = "hs401x",
+		.of_match_table = hs401x_of_match,
+	},
+	.probe	= hs401x_probe,
+	.id_table	= hs401x_id,
+};
+
+module_i2c_driver(hs401x_driver);
+
+MODULE_AUTHOR("Eilin Li <eilin_li@wiwynn.com>");
+MODULE_DESCRIPTION("Renesas Instruments HS401x relative humidity and temperature sensor driver");
+MODULE_LICENSE("GPL");
-- 
2.47.1

